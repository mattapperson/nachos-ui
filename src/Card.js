import React from 'react'
import PropTypes from 'prop-types'
import { View, Image } from 'react-native'
import P from './typography/P'
import { withTheme } from './Theme'

const defaultTheme = {
  CARD_BORDER_RADIUS: 5,
  CARD_BORDER_COLOR: '#eee',
  CARD_MIN_WIDTH: 70,
  CARD_MIN_HEIGHT: 200,
  CARD_FOOTER_HEIGHT: 40,
  CARD_FOOTER_FONT_SIZE: 14,
}

const Card = (props) => {
  const { width, height, theme } = props

  let bodyComponent
  if (props.bodyContent) {
    bodyComponent = props.bodyContent
  }

  let imageComponent
  if (props.image) {
    imageComponent = (
      <View style={theme.imageWrapper}>
        <Image
          source={{ uri: props.image }}
          style={[theme.image, props.imageStyle]}
        />
      </View>
    )
  }

  let footerComponent
  if (props.footerContent) {
    const isString = typeof props.footerContent === 'string'
    let { footerContent } = props
    if (isString) {
      footerContent = (
        <P style={theme.footerText} numberOfLines={1}>
          {footerContent}
        </P>
      )
    }

    footerComponent = (
      <View style={[theme.footer, props.footerStyle]}>
        {footerContent}
      </View>
    )
  }

  return (
    <View style={[theme.card, { width, height }, props.style]}>
      <View style={theme.body}>
        {imageComponent}
        {bodyComponent}
      </View>
      {footerComponent}
    </View>
  )
}

Card.defaultStyle = (theme = defaultTheme) => {
  return {
    card: {
      borderWidth: 1,
      borderRadius: theme.CARD_BORDER_RADIUS,
      borderColor: theme.CARD_BORDER_COLOR,
      flexDirection: 'column',
      alignItems: 'stretch',
      alignSelf: 'stretch',
      minWidth: theme.CARD_MIN_WIDTH,
      minHeight: theme.CARD_MIN_HEIGHT,
      overflow: 'hidden',
    },
    body: { position: 'relative', alignSelf: 'stretch', flex: 1 },
    imageWrapper: {
      position: 'absolute',
      top: 0,
      right: 0,
      bottom: 0,
      left: 0,
    },
    image: { flex: 1, resizeMode: 'cover' },
    footer: {
      overflow: 'hidden',
      borderTopWidth: 1,
      borderColor: '#efefef',
      height: theme.CARD_FOOTER_HEIGHT,
      paddingHorizontal: 8,
      alignSelf: 'stretch',
    },
    footerText: {
      fontSize: theme.CARD_FOOTER_FONT_SIZE,
      lineHeight: theme.CARD_FOOTER_HEIGHT,
      paddingTop: 0,
      paddingBottom: 0,
    },
  }
}

Card.propTypes = {
  width: PropTypes.number,
  height: PropTypes.number,
  style: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
  bodyContent: PropTypes.node,
  image: PropTypes.string,
  imageStyle: PropTypes.oneOfType([
    PropTypes.object,
    PropTypes.array,
  ]),
  footerContent: PropTypes.oneOfType([
    PropTypes.node,
    PropTypes.string,
  ]),
  footerStyle: PropTypes.oneOfType([
    PropTypes.object,
    PropTypes.array,
  ]),
  theme: PropTypes.object,
}

export default withTheme('Card', Card)
