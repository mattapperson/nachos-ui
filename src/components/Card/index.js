import React from "react";
import PropTypes from "prop-types";
import { View, Image } from "react-native";
import P from "../typography/P";
import { withTheme } from "../Theme";

const Card = props => {
  const { width, height, theme } = props;

  let bodyComponent;
  if (props.bodyContent) {
    bodyComponent = props.bodyContent;
  }

  let imageComponent;
  if (props.image) {
    imageComponent = (
      <View style={theme.imageWrapper}>
        <Image
          source={{ uri: props.image }}
          style={[theme.image, props.imageStyle]}
        />
      </View>
    );
  }

  let footerComponent;
  if (props.footerContent) {
    const isString = typeof props.footerContent === "string";
    let { footerContent } = props;
    if (isString) {
      footerContent = (
        <P style={theme.footerText} numberOfLines={1}>
          {footerContent}
        </P>
      );
    }

    footerComponent = (
      <View style={[theme.footer, props.footerStyle]}>{footerContent}</View>
    );
  }

  return (
    <View style={[theme.card, { width, height }, props.style]}>
      <View style={theme.body}>
        {imageComponent}
        {bodyComponent}
      </View>
      {footerComponent}
    </View>
  );
};

Card.themeConfig = {
  style: {
    card: {
      borderWidth: 1,
      borderRadius: 5,
      borderColor: "#eee",
      flexDirection: "column",
      alignItems: "stretch",
      alignSelf: "stretch",
      minWidth: 70,
      minHeight: 200,
      overflow: "hidden"
    },
    body: { position: "relative", alignSelf: "stretch", flex: 1 },
    imageWrapper: {
      position: "absolute",
      top: 0,
      right: 0,
      bottom: 0,
      left: 0
    },
    image: { flex: 1, resizeMode: "cover" },
    footer: {
      overflow: "hidden",
      borderTopWidth: 1,
      borderColor: "#efefef",
      height: 40,
      paddingHorizontal: 8,
      alignSelf: "stretch"
    },
    footerText: {
      paddingTop: 10,
      paddingBottom: 0
    }
  }
};

Card.propTypes = {
  width: PropTypes.number,
  height: PropTypes.number,
  style: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
  bodyContent: PropTypes.node,
  image: PropTypes.string,
  imageStyle: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
  footerContent: PropTypes.oneOfType([PropTypes.node, PropTypes.string]),
  footerStyle: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
  theme: PropTypes.object
};

export default withTheme("Card", Card);
