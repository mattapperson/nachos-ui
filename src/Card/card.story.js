import { storiesOf } from "../storybook";
import { withInfo } from "../storybook";
import React from "react";
import Card from "./index";
import { View } from "react-native";

const Example = () => {
  const cardStyle = { margin: 15, width: 280 };

  return (
    <View style={{ flex: 1, flexDirection: "column" }}>
      <Card
        footerContent="The avocado is a tree that is native to Mexico"
        image="https://upx.cz/BsN"
        style={cardStyle}
      />
      <Card
        footerContent="Agaves are a type of succulent plant from Mexico"
        image="https://upx.cz/BsD"
        style={cardStyle}
      />
    </View>
  );
};

storiesOf("Card", module).add("default", withInfo()(Example));

export default Example;
