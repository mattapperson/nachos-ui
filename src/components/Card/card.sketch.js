import React from "react";
import Card from "./index";

const cardStyle = { margin: 15, width: 280 };
export const components = [
  {
    Component: () => (
      <Card
        footerContent="The avocado is a tree that is native to Mexico"
        image="https://upx.cz/BsN"
        style={cardStyle}
      />
    ),
    name: "Card"
  },
  {
    Component: () => (
      <Card
        footerContent="Agaves are a type of succulent plant from Mexico"
        image="https://upx.cz/BsD"
        style={cardStyle}
      />
    ),
    name: "Card/other"
  }
];
