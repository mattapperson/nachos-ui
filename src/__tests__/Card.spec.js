import { View, Text } from 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'

import Card from '../Card'

it('should render correctly', () => {
  const component = renderer.create(
    <Card footerContent='Footer content' image='image.jpg' />
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

it('should render custom footerContent', () => {

  const component = renderer.create(
    <Card
      footerContent={
        (
          <View style={{ position: 'absolute' }}>
            <Text>Custom footer content</Text>
          </View>
        )
      }
    />
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
