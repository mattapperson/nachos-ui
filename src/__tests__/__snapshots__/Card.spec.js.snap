// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="rn-alignItems-1oszu61 rn-alignSelf-1pz39u2 rn-borderTopColor-6bk7wz rn-borderRightColor-1a82egj rn-borderBottomColor-1jl9v37 rn-borderLeftColor-tn3if rn-borderTopLeftRadius-1as3g83 rn-borderTopRightRadius-dfv94e rn-borderBottomRightRadius-qpmf2f rn-borderBottomLeftRadius-re1h2s rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-5kkj8d rn-borderRightWidth-13l2t4g rn-borderBottomWidth-qklmqi rn-borderLeftWidth-1ljd8xs rn-boxSizing-deolkf rn-display-6koalj rn-flexShrink-1pxmb3b rn-flexBasis-7vfszb rn-flexDirection-eqz5dr rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ql8eny rn-minWidth-1m5eot9 rn-overflowX-11yh6sk rn-overflowY-buy8e9 rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-zIndex-1lgpqti"
>
  <div
    className="rn-alignItems-1oszu61 rn-alignSelf-1pz39u2 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-6koalj rn-flex-13awgt0 rn-flexGrow-1m1wadx rn-flexShrink-1awmn5t rn-flexDirection-eqz5dr rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-zIndex-1lgpqti"
  >
    <div
      className="rn-alignItems-1oszu61 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-bottom-1p0dtai rn-boxSizing-deolkf rn-display-6koalj rn-flexShrink-1pxmb3b rn-flexBasis-7vfszb rn-flexDirection-eqz5dr rn-left-1d2f490 rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-u8s1d rn-right-zchlnj rn-top-ipm5af rn-zIndex-1lgpqti"
    >
      <div
        className="rn-alignItems-1oszu61 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-6koalj rn-flex-13awgt0 rn-flexGrow-1m1wadx rn-flexShrink-1awmn5t rn-flexDirection-eqz5dr rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-zIndex-1lgpqti"
        style={
          Object {
            "backgroundColor": "transparent",
            "backgroundPosition": "center",
            "backgroundRepeat": "no-repeat",
            "backgroundSize": "cover",
          }
        }
      />
    </div>
  </div>
  <div
    className="rn-alignItems-1oszu61 rn-alignSelf-1pz39u2 rn-borderTopColor-1de8lnu rn-borderRightColor-sk4zl rn-borderBottomColor-1nc7mkd rn-borderLeftColor-nyitwm rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-5kkj8d rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-6koalj rn-flexShrink-1pxmb3b rn-flexBasis-7vfszb rn-flexDirection-eqz5dr rn-height-eu3ka rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-overflowX-11yh6sk rn-overflowY-buy8e9 rn-paddingTop-wk8lta rn-paddingRight-1pyaxff rn-paddingBottom-1mdbw0j rn-paddingLeft-1m04atk rn-position-bnwqim rn-zIndex-1lgpqti"
  >
    <div
      className="rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-color-15b02fe rn-display-1471scf rn-fontFamily-14xgk7a rn-fontSize-1b43r93 rn-fontStyle-13uqrnb rn-fontVariant-ebii48 rn-fontWeight-1it3c9n rn-lineHeight-t9a87b rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-maxWidth-dnmrzs rn-overflowX-11yh6sk rn-overflowY-buy8e9 rn-paddingRight-9aemit rn-paddingLeft-gy4na3 rn-paddingBottom-1mdbw0j rn-paddingTop-m611by rn-textAlign-fdjqy7 rn-textDecoration-bauka4 rn-textOverflow-1udbk01 rn-whiteSpace-3s2u2q rn-wordWrap-qvutc0"
      dir="auto"
    >
      Footer content
    </div>
  </div>
</div>
`;

exports[`should render custom footerContent 1`] = `
<div
  className="rn-alignItems-1oszu61 rn-alignSelf-1pz39u2 rn-borderTopColor-6bk7wz rn-borderRightColor-1a82egj rn-borderBottomColor-1jl9v37 rn-borderLeftColor-tn3if rn-borderTopLeftRadius-1as3g83 rn-borderTopRightRadius-dfv94e rn-borderBottomRightRadius-qpmf2f rn-borderBottomLeftRadius-re1h2s rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-5kkj8d rn-borderRightWidth-13l2t4g rn-borderBottomWidth-qklmqi rn-borderLeftWidth-1ljd8xs rn-boxSizing-deolkf rn-display-6koalj rn-flexShrink-1pxmb3b rn-flexBasis-7vfszb rn-flexDirection-eqz5dr rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ql8eny rn-minWidth-1m5eot9 rn-overflowX-11yh6sk rn-overflowY-buy8e9 rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-zIndex-1lgpqti"
>
  <div
    className="rn-alignItems-1oszu61 rn-alignSelf-1pz39u2 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-6koalj rn-flex-13awgt0 rn-flexGrow-1m1wadx rn-flexShrink-1awmn5t rn-flexDirection-eqz5dr rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-zIndex-1lgpqti"
  />
  <div
    className="rn-alignItems-1oszu61 rn-alignSelf-1pz39u2 rn-borderTopColor-1de8lnu rn-borderRightColor-sk4zl rn-borderBottomColor-1nc7mkd rn-borderLeftColor-nyitwm rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-5kkj8d rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-6koalj rn-flexShrink-1pxmb3b rn-flexBasis-7vfszb rn-flexDirection-eqz5dr rn-height-eu3ka rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-overflowX-11yh6sk rn-overflowY-buy8e9 rn-paddingTop-wk8lta rn-paddingRight-1pyaxff rn-paddingBottom-1mdbw0j rn-paddingLeft-1m04atk rn-position-bnwqim rn-zIndex-1lgpqti"
  >
    <div
      className="rn-alignItems-1oszu61 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-6koalj rn-flexShrink-1pxmb3b rn-flexBasis-7vfszb rn-flexDirection-eqz5dr rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-u8s1d rn-zIndex-1lgpqti"
    >
      <div
        className="rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-color-homxoj rn-display-1471scf rn-fontFamily-14xgk7a rn-fontSize-1b43r93 rn-fontStyle-o11vmf rn-fontVariant-ebii48 rn-fontWeight-gul640 rn-lineHeight-t9a87b rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-textDecoration-bauka4 rn-whiteSpace-q42fyq rn-wordWrap-qvutc0"
        dir="auto"
      >
        Custom footer content
      </div>
    </div>
  </div>
</div>
`;
